tests/test_type_adapter.py

```diff
@@ -371,3 +473,108 @@ def test_eval_type_backport():
     assert exc_info.value.errors(include_url=False) == [
         {'type': 'list_type', 'loc': (), 'msg': 'Input should be a valid list', 'input': 'not a list'}
     ]
+
+
+def defer_build_test_models(config: ConfigDict) -> List[Any]:
+    class Model(BaseModel):
+        model_config = config
+        x: int
+
+    class SubModel(Model):
+        y: Optional[int] = None
+
+    @pydantic_dataclass(config=config)
+    class DataClassModel:
+        x: int
+
+    @pydantic_dataclass
+    class SubDataClassModel(DataClassModel):
+        y: Optional[int] = None
+
+    class TypedDictModel(TypedDict):
+        __pydantic_config__ = config  # type: ignore
+        x: int
+
+    models = [
+        Model,
+        SubModel,
+        create_model('DynamicModel', __base__=Model),
+        create_model('DynamicSubModel', __base__=SubModel),
+        DataClassModel,
+        SubDataClassModel,
+        TypedDictModel,
+        Dict[str, int],
+    ]
+    return [
+        *models,
+        # FastAPI heavily uses Annotated so test that as well
+        *[Annotated[model, Field(title='abc')] for model in models],
+    ]
+
+
+CONFIGS = [
+    ConfigDict(defer_build=False, _defer_build_mode=('model',)),
+    ConfigDict(defer_build=False, _defer_build_mode=DEFER_ENABLE_MODE),
+    ConfigDict(defer_build=True, _defer_build_mode=('model',)),
+    ConfigDict(defer_build=True, _defer_build_mode=DEFER_ENABLE_MODE),
+]
+MODELS_CONFIGS: List[Tuple[Any, ConfigDict]] = [
+    (model, config) for config in CONFIGS for model in defer_build_test_models(config)
+]
+
+
+@pytest.mark.parametrize('model, config', MODELS_CONFIGS)
```

Comment:
Could we further consolidate these last 3 tests based on their similarities (especially the first half of the logic)?

Reply:
Nice